flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentInternal.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.api.internal;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.api.TableEnvironment;
+import org.apache.flink.table.api.TableResult;
+import org.apache.flink.table.catalog.CatalogManager;
+import org.apache.flink.table.delegation.Parser;
+import org.apache.flink.table.operations.ModifyOperation;
+
+import java.util.List;
+
+/**
+ * An internal interface of {@link TableEnvironment}
+ * that defines extended methods used for {@link TableImpl}.
+ */
+@Internal
+public interface TableEnvironmentInternal extends TableEnvironment {
```

Comment:
How about adding all the methods to `TableEnvironment` and mark them as `@Internal`?

Reply:
`@Internal` is weak constraint, If users do not look into source code, he/she even does not find whether a method is internal method. These internal methods may be unstable, I trend to we do not expose them into public interface.

Reply:
@twalthr what's your suggestion to this?

Reply:
No, we should definitely keep the API clean. However, I still not understand the need for `TableEnvironmentInternal `. We have so many table environments already in the API classpath which is quite confusion. Why can't a `TableImpl` just cast to `TableEnvironmentImpl` internally.